File: m/problem5.m
clc; close all; clear variables;

function [c, iter, err] = fixed_point_method(g, x_0, tol, N)
    if ~isa(g, 'function_handle')
        fprintf('g(x) must be callable (e.g., a function).');
        return;
    end
    if ~isnumeric(x_0) || ~isnumeric(tol)
        fprintf('x_0 and tol must be numbers.');
        return;
    end
    if ~isnumeric(N) || N < 1 || mod(N, 1) ~= 0
        fprintf('N must be a natural number.');
        return;
    end
    for i = 1:N
        fprintf(x_0);
        c = g(x_0);
        err = abs(c - x_0);
        if err < tol
            iter = i;
            return;
        end
        x_0 = c;
    end

    fprintf('Maximum iterations exceeded');
    iter = i;
    err = abs(c - x_0);
    return;
end

File: py/problem5.py
def fixed_point_method(g: Callable, x_0: float = 0, tol: float = 1e-5, N: int = 100) -> (float, int, float):
    if(not callable(g)):
        print("g(x) must be callable (e.g., a function).")
        return
    numbers = [int, float]
    if(type(x_0) not in numbers or type(tol) not in numbers):
        print("x_0 and tol must be numbers.")
        return
    if(type(N) != int or N < 1):
        print("N must be a natural number")
        return

    for i in range(N):
        print(x_0)
        c = g(x_0)
        err = abs(c - x_0)
        if(abs(c - x_0) < tol):
            return(c, i, err)
        x_0 = c

    print("Maximum iterations exceeded")
    return(x, i, err)
#
#
#
#
#
#
#
#
#